### ISSUE TITLE
Strage lock corruption on macOS when used from rust and rayon

### ISSUE THREAD

**Issue by lu-zero** [amzn/hawktracer#64]
Follow up from https://github.com/AlexEne/rust_hawktracer/issues/25

Reproduction steps:

```
git clone https://github.com/xiph/rav1e
cd rav1e
wget http://ultravideo.cs.tut.fi/video/Bosphorus_1920x1080_120fps_420_8bit_YUV_Y4M.7z
7z x Bosphorus_1920x1080_120fps_420_8bit_YUV_Y4M.7z
cargo run --release --features=tracing -- ./Bosphorus_1920x1080_120fps_420_8bit_YUV.y4m --tiles 8 -o /dev/null -l 30
```

It randomly crashes on close. You may want to use content with smaller resolution.

**Comment by loganek** [comment]
Thanks for the report. I looked at rust hawktracer code, and I think there might be a bug in the bindings.
According to the example, listener gets dropped at the end of the main function, but there might be some other threads running and using the listener. So the proper fix would be to detach the listener from the timeline just before dropping it.
I'll post the comment on the other ticket and help @AlexEne fixing the bug.
I'll also leave the ticket open as long as https://github.com/AlexEne/rust_hawktracer/issues/25 is open just in case there might be some changes required in the hawktracer itself.

**Comment by loganek** [comment]
@lu-zero sorry for the late response. Could you recall what version of rust_hawktracer did you use to reproduce the problem?

**Comment by lu-zero** [comment]
I managed to reproduce it again with the `0.6.0` release today. It took me few tries to get there =/

**Comment by loganek** [comment]
@lu-zero thanks for the information, and I totally understand your frustration. I was just trying to reproduce the failure by following your reproduction steps, but the code failed to compile:
```
   Compiling rav1e v0.2.0 (/home/ANT.AMAZON.COM/mkolny/repos/rav1e)
error: failed to run custom build command for `rav1e v0.2.0 (/home/ANT.AMAZON.COM/mkolny/repos/rav1e)`

Caused by:
  process didn't exit successfully: `/home/ANT.AMAZON.COM/mkolny/repos/rav1e/target/release/build/rav1e-953fcc3679d2c952/build-script-build` (exit code: 101)
--- stdout
cargo:rustc-cfg=nasm_x86_64

--- stderr
thread 'main' panicked at 'This version of NASM is too old: No such file or directory (os error 2)', crates/nasm_rs/src/lib.rs:323:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed
```
I'd be grateful if you could help me compiling the project so I could reproduce the issue and collect all the data I need for fixing it.
Many thanks!

**Comment by loganek** [comment]
@lu-zero I'm very sorry to hear that. I've tried to reproduce the bug by following your steps (I executed it 200 times) and didn't manage to reproduce the problem. Did you by any chance manage to get core dumps of any of those failures?
Also, I looked at your code and have a question that could help diagnose the issue; does `_listener` variable in `rav1e.rs` file outlives all the other threads except the main thread? Or are there any threads in your application that finish after `main()` function? (HawkTracer does support threads that live longer than `main()`, but just wanted to check if it's worth going that path in my investigation)

Many thanks for your understanding and the support.

**Comment by loganek** [comment]
@lu-zero I found a bug based on analyzing the code and managed to easily reproduce it by adding a few sleeps in the code. The fix is already in master (https://github.com/amzn/hawktracer/commit/ea9780ddcab8c7cab08e32f95f2c9a894a270b25) and in the [latest release (0.10.0)](https://github.com/amzn/hawktracer/releases/tag/v0.10.0).
Unfortunately, it also requires some changes in rust bindings, so I've already made a pull request (https://github.com/AlexEne/rust_hawktracer_sys/pull/10). I'll work with @AlexEne to have it included in the next rust_hawktracer release.
I'm very that fixing the issue took so long, I hope it didn't cause a lot of inconvenience.
